Use dedicated pointer types in syscall layer. NFC

kleisauke opened the pull request:
Split out from #13007.

sbc100 reviewed (approved):
If the tests pass then LGTM.

IIRC there was a reason we did it this way when porting the musl syscall layer, but I can't remember exactly what that was.. so lets just go with this if the tests pass.

sbc100 commented on system/lib/libc/emscripten_syscall_stubs.c:
```diff
@@ -82,46 +82,46 @@ weak int __syscall_sync() {
   return g_ppid;
 }
 
 weak int __syscall_linkat(int olddirfd, intptr_t oldpath, int newdirfd, intptr_t newpath, int flags) {
```
Should we also use pointer types here for oldpath/newpath?  and elsewhere?

sbc100 commented on system/lib/libc/emscripten_syscall_stubs.c:
```diff
@@ -82,46 +82,46 @@ weak int __syscall_sync() {
 
-weak int __syscall_getgroups32(int size, intptr_t list) {
-  if (size < 1) {
+weak int __syscall_getgroups32(int count, intptr_t list) {
```
Should list be `gid_t*` (avoiding the case below?)

kleisauke committed:
Use dedicated pointer types in syscall layer. NFC

sbc100 reviewed (commented):
Can we avoid the all the changes/additions to the sanitizers?   This change seems a lot more risky once you add those in?

Also changing the socket syscalls to be varargs seems a little out of scope and perhaps separable.

sbc100 commented on system/lib/libc/musl/arch/emscripten/syscall_arch.h:
```diff
@@ -12,111 +17,114 @@
-int __syscall_chmod(intptr_t path, int mode);
-int __syscall_getpid(void);
+typedef unsigned socklen_t;
+typedef unsigned long nfds_t;
```
Duplicating these definitions here seems a little sad.

kleisauke commented:
Let me mark this as a draft, there are many more changes needed than expected. For example, doing this:
```diff
-int __syscall_openat(int dirfd, intptr_t path, int flags, ...); // mode is optional
+int __syscall_openat(int dirfd, const char *path, int flags, ...); // mode is optional
```

Will fail within the sanitizers (which uses the internal `syscall` function of musl):
https://github.com/emscripten-core/emscripten/blob/b182182747e8a5525ff3282953fb814c8c0c9266/system/lib/compiler-rt/lib/sanitizer_common/sanitizer_linux.cpp#L262
https://github.com/emscripten-core/emscripten/blob/b182182747e8a5525ff3282953fb814c8c0c9266/tools/system_libs.py#L1925-L1928

<details>
  <summary>Details</summary>

```
/home/kleisauke/emscripten/system/lib/compiler-rt/lib/sanitizer_common/sanitizer_syscall_generic.inc:17:24: note: expanded from macro 'SYSCALL'
   17 | # define SYSCALL(name) SYS_ ## name
      |                        ^~~~~~~~~~~~
<scratch space>:132:1: note: expanded from here
  132 | SYS_openat
      | ^~~~~~~~~~
/home/kleisauke/emscripten/cache/sysroot/include/bits/syscall.h:61:21: note: expanded from macro 'SYS_openat'
   61 | #define SYS_openat              __syscall_openat
      |                                 ^~~~~~~~~~~~~~~~
note: (skipping 2 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
/home/kleisauke/emscripten/system/lib/libc/musl/src/internal/syscall.h:65:80: note: expanded from macro '__SYSCALL_DISP'
   65 | #define __SYSCALL_DISP(b,...) __SYSCALL_CONCAT(b,__SYSCALL_NARGS(__VA_ARGS__))(__VA_ARGS__)
      |                                                                                ^~~~~~~~~~~
/home/kleisauke/emscripten/system/lib/libc/musl/src/internal/syscall.h:55:61: note: expanded from macro '__syscall_emscripten3'
   55 | #define __syscall_emscripten3(n,a,b,c) __syscall_emscripten(n,__scc(a),__scc(b),__scc(c))
      |                                                             ^
/home/kleisauke/emscripten/system/lib/libc/musl/src/internal/syscall.h:51:38: note: expanded from macro '__syscall_emscripten'
   51 | #define __syscall_emscripten(n, ...) n(__VA_ARGS__)
      |                                      ^
/home/kleisauke/emscripten/cache/sysroot/include/syscall_arch.h:93:5: note: candidate function not viable: no known conversion from 'uptr' (aka 'unsigned long') to 'const char *' for 2nd argument
   93 | int __syscall_openat(int dirfd, const char *path, int flags, ...); // mode is optional
      |     ^
/home/kleisauke/emscripten/system/lib/compiler-rt/lib/sanitizer_common/sanitizer_linux.cpp:270:27: error: no matching function for call to '__syscall_openat'
  270 |   return internal_syscall(SYSCALL(openat), AT_FDCWD, (uptr)filename, flags,

```
</details>

The solution is to rework the sanitizers to avoid musl's syscall layer, similar to the NetBSD and macOS implementations.
https://github.com/emscripten-core/emscripten/blob/b182182747e8a5525ff3282953fb814c8c0c9266/system/lib/compiler-rt/lib/sanitizer_common/sanitizer_mac.cpp#L165-L167

sbc100 commented:
Yeah I think this is one reason things are the way they today.. to allow the SYSCALL stuff to work.

Perhaps we could land the initial/smaller version of this change that doesn't cause some much churn?

kleisauke commented:
I've just split the compiler-rt changes to PR #24655.

kleisauke committed:
Fix build failures

By reverting musl commit 5850546e9669f793aab61dfc7c4f2c1ff35c4b29.

kleisauke commented:
This should be in good shape now, but it's still very difficult to review. I'll probably split it into 3 PRs instead (i.e. one for commit a66926dbf49910399bcb1dc0b1c52823d0319c1d, 28ff6378f93d7123f1cc4f083b41e78b915167c2 and c5f7ca94e1bf06f6cf7037551554c15979c31ab7).

sbc100 commented on system/lib/libc/emscripten_syscall_stubs.c:
```diff
@@ -82,105 +74,97 @@ weak int __syscall_sync() {
-  u->stime_tv_sec = 3;
-  u->stime_tv_usec = 4;
+  usage->ru_utime = (struct timeval)
+    { .tv_sec = 1, .tv_usec = 2 };
```
This looks like it could fit on one line?

(do you find this syntax preferable?)

kleisauke replied:
I was trying to keep the syntax in-line with musl, see e.g.:
https://github.com/emscripten-core/emscripten/blob/22913506379102da2f1f51acf9b492f419b2e413/system/lib/libc/musl/src/misc/getrusage.c#L29-L32

sbc100 commented on system/lib/libc/emscripten_syscall_stubs.c:
```diff
@@ -221,42 +205,45 @@ weak int __syscall_munlockall() {
 }
 
-weak int __syscall_prlimit64(int pid, int resource, intptr_t new_limit, intptr_t old_limit) {
+weak int __syscall_prlimit64(pid_t pid, int resource, const struct rlimit *new_limit, struct rlimit *old_limit) {
```
Maybe rename the param to `old` to minimize the diff here?

kleisauke replied:
I'm not sure, since that would make grepping more difficult (though only if someone tries to `grep` the full syscall including all its arguments). Also, I would need to rename `new_limit` to `new` for consistency, which is a reserved C++ keyword.

The best approach I can think of is to split this into two commits: one that only updates the function signatures, and another that fixes the resulting build errors and removes the redundant casts. Let me know if you'd prefer that. :)

sbc100 commented on system/lib/wasmfs/syscalls.cpp:
```diff
@@ -346,67 +346,69 @@ backend_t wasmfs_get_backend_by_path(const char* path) {
-int __syscall_newfstatat(int dirfd, intptr_t path, intptr_t buf, int flags) {
+int __syscall_newfstatat(int dirfd,
+                         const char* path,
+                         struct stat* buf,
```
Maybe rename this to buffer to minimize the diff here?

sbc100 commented on system/lib/libc/musl/src/linux/wait4.c:
```diff
@@ -7,6 +7,9 @@
 {
+#ifdef __EMSCRIPTEN__ // XXX Emscripten revert musl commit 5850546e9669f793aab61dfc7c4f2c1ff35c4b29
+	return syscall(SYS_wait4, pid, status, options, ru);
+#else
```
I'm curious about this and that rusage change?  They seem a little different to rest of this change.  Should we land them first/separately?

kleisauke replied:
This is part of commit 380661414f6a87b35fdd533b30087ff0031e72d3, which should probably land first (hence why I haven't removed the draft status yet).

sbc100 reviewed (commented):
This is looking really good now.  Can we remove the draft status?

kleisauke commented:
Note to self: perhaps drop commit b92f1f70e256ba6ccbd0ddb7ec2fb28ed4d84e06 (but keep the `__syscall_recvmmsg` and `__syscall_sendmmsg` changes), as marshalling varargs may be more expensive on wasm64 than passing a couple of integer arguments.

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (3) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_files_js_fs.json: 18247 => 18247 [+0 bytes / +0.00%]
codesize/test_codesize_files_wasmfs.json: 63892 => 63892 [+0 bytes / +0.00%]
codesize/test_codesize_hello_dylink_all.json: 855818 => 855594 [-224 bytes / -0.03%]

Average change: -0.01% (-0.03% - +0.00%)
```

kleisauke commented:
This has been a long journey. I'm marking this as ready for review, but it may still be difficult to review (and ideally commit 3838302569d2523913c380883c1aa9f596b4871d should be split out as well).

On the positive side, this appears to save 224 bytes in `test_codesize_hello_dylink_all.json`.

sbc100 commented on system/lib/libc/emscripten_internal.h:
```diff
@@ -84,9 +84,9 @@ int _mmap_js(size_t length,
              void** addr);
 int _munmap_js(
-  intptr_t addr, size_t length, int prot, int flags, int fd, off_t offset);
+  void *addr, size_t length, int prot, int flags, int fd, off_t offset);
```
Do these fit on one line now?

kleisauke replied:
Ah, I only ran `clang-format` on the C++ files, let me check.

kleisauke replied:
Unfortunately, no, these don't fit on one line. `git-clang-format` (as opposed to `clang-format`, which changes unrelated lines) does seem to alter pointer alignment, see commit e367450bdcd4fb8a0c8772a73cbcc1f643fe4ecf.

sbc100 reviewed (commented):
Nice!

Thanks for the persistence on this.

kleisauke committed:
`git-clang-format` on `emscripten_internal.h`

kleisauke committed:
Merge branch 'main' into correct-types

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_hello_dylink_all.json: 856149 => 855925 [-224 bytes / -0.03%]

Average change: -0.03% (-0.03% - -0.03%)
```

kleisauke committed:
Automatic rebaseline of native signature info. NFC

This is an automatic change generated by tools/maint/gen_native_sig_info.py.
